File: frontend/src/lib/api/ops.ts
// src/lib/api/ops.ts — Operations & facilities API
import { fetchClient } from './client';

export interface FacilityAsset {
  id: string;
  name: string;
  type: string;
  location?: string;
  status: string;
  created_at: string;
}

export interface AssetBooking {
  id: string;
  asset_id: string;
  employee_id: string;
  start_time: string;
  end_time: string;
  status: string;
  created_at: string;
}

export interface CalendarBooking {
  id: string;
  asset_id: string;
  asset_name?: string;
  asset_type?: string;
  employee_name?: string;
  start_time: string;
  end_time: string;
  status: string;
}

export interface VisitorLog {
  id: string;
  visitor_name: string;
  company?: string;
  host_id: string;
  expected_arrival: string;
  check_in_time?: string;
  check_out_time?: string;
  status: string;
  created_at: string;
}

export interface MaintenanceRequest {
  id: string;
  asset_id?: string;
  title: string;
  description?: string;
  priority: string;
  status: string;
  reported_by_id: string;
  assigned_to_id?: string;
  resolution_notes?: string;
  created_at: string;
  resolved_at?: string;
}

export interface MaintenanceStats {
  total: number;
  reported: number;
  in_progress: number;
  resolved: number;
  closed: number;
  low: number;
  medium: number;
  high: number;
  critical: number;
}

export const OpsAPI = {
  getAssets: (type?: string) => fetchClient(`/ops/assets${type ? `?type=${type}` : ""}`),
  createAsset: (data: Partial<FacilityAsset>) => fetchClient("/ops/assets", { method: "POST", body: JSON.stringify(data) }),
  getBookings: (employeeId?: string) => fetchClient(`/ops/bookings${employeeId ? `?employee_id=${employeeId}` : ""}`),
  createBooking: (data: Partial<AssetBooking>) => fetchClient("/ops/bookings", { method: "POST", body: JSON.stringify(data) }),
  getVisitors: (hostId?: string) => fetchClient(`/ops/visitors${hostId ? `?host_id=${hostId}` : ""}`),
  createVisitor: (data: Partial<VisitorLog>) => fetchClient("/ops/visitors", { method: "POST", body: JSON.stringify(data) }),
  updateVisitor: (visitorId: string, data: Partial<VisitorLog>) => fetchClient(`/ops/visitors/${visitorId}`, { method: "PUT", body: JSON.stringify(data) }),
  checkOutVisitor: (visitorId: string) => fetchClient(`/ops/visitors/${visitorId}/check-out`, { method: "PUT" }),
  getCalendarBookings: (start: string, end: string, assetType?: string) =>
    fetchClient(`/ops/bookings/calendar?start=${encodeURIComponent(start)}&end=${encodeURIComponent(end)}${assetType ? `&asset_type=${encodeURIComponent(assetType)}` : ""}`),
  getMaintenanceRequests: (status?: string, priority?: string) => {
    const params = new URLSearchParams();
    if (status) params.set("status", status);
    if (priority) params.set("priority", priority);
    const qs = params.toString();
    return fetchClient(`/ops/maintenance${qs ? `?${qs}` : ""}`);
  },
  createMaintenanceRequest: (data: { title: string; description?: string; asset_id?: string; priority?: string; reported_by_id: string }) =>
    fetchClient("/ops/maintenance", { method: "POST", body: JSON.stringify(data) }),
  updateMaintenanceRequest: (requestId: string, data: Record<string, unknown>) =>
    fetchClient(`/ops/maintenance/${requestId}`, { method: "PUT", body: JSON.stringify(data) }),
  getMaintenanceStats: () => fetchClient("/ops/maintenance/stats"),
};

File: backend/app/schemas/ops.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime

class FacilityAssetBase(BaseModel):
    name: str
    type: str
    location: Optional[str] = None
    status: Optional[str] = "available"

class FacilityAssetCreate(FacilityAssetBase):
    pass

class FacilityAssetResponse(FacilityAssetBase):
    id: str
    created_at: datetime

    class Config:
        from_attributes = True

class AssetBookingBase(BaseModel):
    asset_id: str
    employee_id: str
    start_time: datetime
    end_time: datetime
    status: Optional[str] = "confirmed"

class AssetBookingCreate(AssetBookingBase):
    pass

class AssetBookingResponse(AssetBookingBase):
    id: str
    created_at: datetime

    class Config:
        from_attributes = True

class VisitorLogBase(BaseModel):
    visitor_name: str
    company: Optional[str] = None
    host_id: str
    expected_arrival: datetime
    status: Optional[str] = "expected"

class VisitorLogCreate(VisitorLogBase):
    pass

class VisitorLogUpdate(BaseModel):
    status: Optional[str] = None
    check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None

class VisitorLogResponse(VisitorLogBase):
    id: str
    check_in_time: Optional[datetime] = None
    check_out_time: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True

class CalendarBookingResponse(BaseModel):
    id: str
    asset_id: str
    asset_name: Optional[str] = None
    asset_type: Optional[str] = None
    employee_name: Optional[str] = None
    start_time: datetime
    end_time: datetime
    status: str

    class Config:
        from_attributes = True

class MaintenanceRequestBase(BaseModel):
    asset_id: Optional[str] = None
    title: str
    description: Optional[str] = ""
    priority: Optional[str] = "medium"
    status: Optional[str] = "reported"
    reported_by_id: str

class MaintenanceRequestCreate(MaintenanceRequestBase):
    pass

class MaintenanceRequestUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[str] = None
    status: Optional[str] = None
    assigned_to_id: Optional[str] = None
    resolution_notes: Optional[str] = None

class MaintenanceRequestResponse(MaintenanceRequestBase):
    id: str
    assigned_to_id: Optional[str] = None
    resolution_notes: Optional[str] = None
    created_at: datetime
    resolved_at: Optional[datetime] = None

    class Config:
        from_attributes = True

class MaintenanceStats(BaseModel):
    total: int
    reported: int
    in_progress: int
    resolved: int
    closed: int
    low: int
    medium: int
    high: int
    critical: int
